### ISSUE TITLE
[BUG] stage 3 cannot load the checkpoint when optimizer is not configured

### ISSUE THREAD

**Issue by rohitgr7** [microsoft/DeepSpeed#2449]
**Describe the bug**
Deepspeed stage 3 cannot load the checkpoint when optimizer is not configured during inferencing.

Error: `'DeepSpeedZeRoOffload' object has no attribute 'checkpoint_event_prologue'`

I used a very basic pytorch-lightning script to generate the issue. I am a maintainer at lightning, so feel free to ping me for any information.
```py
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        accelerator='gpu',
        devices=2,
        strategy='deepspeed_stage_3'
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.save_checkpoint('deepspeed_ckpt')
    trainer.test(model, dataloaders=test_data, ckpt_path='deepspeed_ckpt')


if __name__ == "__main__":
    run()

```

**To Reproduce**
1. Run the above script
2. deepspeed==0.7.4, torch==1.12.1+cu116
3. python script_name.py

**Expected behavior**
It should work without any error since the engine is reinitilaized for inference without any optimizer as mentioned here: https://deepspeed.readthedocs.io/en/latest/model-checkpointing.html#deepspeed.DeepSpeedEngine.load_checkpoint

**ds_report output**
Please run `ds_report` to give us details about your setup.

```console
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
sparse_attn ............ [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/rohit/miniconda3/envs/pl/lib/python3.9/site-packages/torch']
torch version .................... 1.12.1+cu116
torch cuda version ............... 11.6
torch hip version ................ None
nvcc version ..................... 11.6
deepspeed install path ........... ['/home/rohit/miniconda3/envs/pl/lib/python3.9/site-packages/deepspeed']
deepspeed info ................... 0.7.4, unknown, unknown
deepspeed wheel compiled w. ...... torch 1.12, cuda 11.6
```

**Screenshots**

**System info (please complete the following information):**
 - OS: [e.g. Ubuntu 18.04]
 - GPU count and types [e.g. two machines with x8 A100s each]
 - (if applicable) what [DeepSpeed-MII](https://github.com/microsoft/deepspeed-mii) version are you using
 - (if applicable) Hugging Face Transformers/Accelerate/etc. versions
 - Python version
 - Any other relevant info about your setup

**Docker context**
Are you using a specific docker image that you can share?

**Additional context**

**Comment by rohitgr7** [comment]
@RezaYazdaniAminabadi

can you take a look? thank you :)

**Comment by tjruwase** [comment]
@rohitgr7, apologies, I think you are running into API issues around inference. Please note that running inference in this manner will enable [ZeRO-Inference](https://www.deepspeed.ai/2022/09/09/zero-inference.html) rather than [DeepSpeed-Inference](https://deepspeed.readthedocs.io/en/latest/inference-engine.html).  Can you please confirm that this is what you expect?

**Comment by tjruwase** [comment]
Irrespective of your answer to the above question, we still need to fix this issue.

**Comment by ShijieZZZZ** [comment]
Hi @rohitgr7, I used your script to try to reproduce the issue. However, the script runs successfully without any error.

`
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NV_LIBNCCL_DEV_PACKAGE=libnccl-dev=2.12.10-1+cuda11.6
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NV_LIBNCCL_DEV_PACKAGE_VERSION=2.12.10-1
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NCCL_VERSION=2.12.10-1
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NV_LIBNCCL_DEV_PACKAGE_NAME=libnccl-dev
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NCCL_ASYNC_ERROR_HANDLING=1
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NCCL_IB_TIMEOUT=20
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NV_LIBNCCL_PACKAGE=libnccl2=2.12.10-1+cuda11.6
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NV_LIBNCCL_PACKAGE_NAME=libnccl2
[2022-11-02 17:46:57,929] [INFO] [launch.py:129:main] 0 NV_LIBNCCL_PACKAGE_VERSION=2.12.10-1
[2022-11-02 17:46:57,930] [INFO] [launch.py:136:main] WORLD INFO DICT: {'master': [0, 1]}
[2022-11-02 17:46:57,930] [INFO] [launch.py:142:main] nnodes=1, num_local_procs=2, node_rank=0
[2022-11-02 17:46:57,930] [INFO] [launch.py:155:main] global_rank_mapping=defaultdict(<class 'list'>, {'master': [0, 1]})
[2022-11-02 17:46:57,930] [INFO] [launch.py:156:main] dist_world_size=2
[2022-11-02 17:46:57,930] [INFO] [launch.py:158:main] Setting CUDA_VISIBLE_DEVICES=0,1
[2022-11-02 17:47:09,711] [WARNING] [engine.py:1152:_configure_optimizer] **** You are using ZeRO with an untested optimizer, proceed with caution *****
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
Emitting ninja build file /home/torch_extensions/py39_cu116/utils/build.ninja...
Building extension module utils...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
ninja: no work to do.
Loading extension module utils...
Time to load utils op: 0.4472646713256836 seconds
Loading extension module utils...
Time to load utils op: 0.3054385185241699 seconds
Parameter Offload: Total persistent parameters: 66 in 2 params
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0010385513305664062 seconds
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0006890296936035156 seconds

Training: 0it [00:00, ?it/s]
Training:   0%|          | 0/2 [00:00<?, ?it/s]
Epoch 0:   0%|          | 0/2 [00:00<?, ?it/s] 
Epoch 0:  50%|█████     | 1/2 [00:00<00:00,  1.24it/s]
Epoch 0:  50%|█████     | 1/2 [00:00<00:00,  1.24it/s, loss=0.53, v_num=19]

Validation: 0it [00:00, ?it/s][A

Validation:   0%|          | 0/1 [00:00<?, ?it/s][A

Validation DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s][A

Validation DataLoader 0: 100%|██████████| 1/1 [00:00<00:00, 275.69it/s][A
Epoch 0: 100%|██████████| 2/2 [00:00<00:00,  2.47it/s, loss=0.53, v_num=19]
Epoch 0: 100%|██████████| 2/2 [00:00<00:00,  2.44it/s, loss=0.53, v_num=19]

                                                                       [A
Epoch 0: 100%|██████████| 2/2 [00:00<00:00,  2.44it/s, loss=0.53, v_num=19]
Epoch 0: 100%|██████████| 2/2 [00:00<00:00,  2.37it/s, loss=0.53, v_num=19]
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0011129379272460938 seconds
[2022-11-02 17:47:12,207] [WARNING] [engine.py:1152:_configure_optimizer] **** You are using ZeRO with an untested optimizer, proceed with caution *****
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0008959770202636719 seconds
Parameter Offload: Total persistent parameters: 66 in 2 params
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0006341934204101562 seconds
Using /home/torch_extensions/py39_cu116 as PyTorch extensions root...
No modifications detected for re-loaded extension module utils, skipping build step...
Loading extension module utils...
Time to load utils op: 0.0009706020355224609 seconds

Testing: 0it [00:00, ?it/s]
Testing:   0%|          | 0/1 [00:00<?, ?it/s]
Testing DataLoader 0:   0%|          | 0/1 [00:00<?, ?it/s]
Testing DataLoader 0: 100%|██████████| 1/1 [00:00<00:00, 100.07it/s]
Testing DataLoader 0: 100%|██████████| 1/1 [00:00<00:00, 62.96it/s] 
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       Test metric             DataLoader 0
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
        test_loss           0.33501216769218445
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
[2022-11-02 17:47:13,965] [INFO] [launch.py:318:main] Process 56543 exits successfully.
[2022-11-02 17:47:13,965] [INFO] [launch.py:318:main] Process 56542 exits successfully.

`

**Comment by rohitgr7** [comment]
hi @ShijieZZZZ, which pytorch-lightning version did you try it with? v1.8?

**Comment by ShijieZZZZ** [comment]
Hi @rohitgr7, pytorch-lightning version = "1.7.7"

**Comment by ShijieZZZZ** [comment]
Observed reported issue with 1.8.0. Investigating.

**Comment by ShijieZZZZ** [comment]
Hi @rohitgr7, if you want to do inferencing with load_checkpoint in zero_stage3, you need to pass an optimizer when initializing deepspeed engine (in _initialize_deepspeed_inference()). Unfortunately, this is required now.

**Comment by jeffra** [comment]
@rohitgr7, please re-open if you're still having issues here.

**Comment by Zehui127** [comment]
What is the solution in this case? How should we modify the code above to make it work?

**Comment by ShijieZZZZ** [comment]
Hi @Zehui127, it needs to specify optimizer [here](https://github.com/Lightning-AI/lightning/blob/0fee28409b00ed8bad9c3ba52477292c9cf4687a/src/lightning/pytorch/strategies/deepspeed.py#:~:text=%3Dmodel%2C-,optimizer%3DNone%2C,-lr_scheduler%3DNone). also check this [PR](https://github.com/Lightning-AI/lightning/commit/0a729f6da10025942d2b369d2a62dff7f150d2ed).

**Comment by Zehui127** [comment]
Thanks for your help @ShijieZZZZ . BTW, I'm curious why this change is introduced? Manually pass in an optimiser during the inference time doesn't seem to be intuitive for the user.

**Comment by ShijieZZZZ** [comment]
Agree. But it seems need non-trivial efforts to completely address.

**Comment by RoyJames** [comment]
This is extremely counter-intuitive and I haven't figured out a way to do it properly purely from a saved checkpoint. Hoping this can be fixed.

**Comment by tjruwase** [comment]
@rohitgr7 and @RoyJames, can you please help test the new fix which avoids the counter-intuitive requirement of an optimizer. Thanks!

**Comment by tjruwase** [comment]
@rohitgr7, thanks for the simple repro code. I extended as below, and the new fix works on it. Since you are lightning maintainer, I will appreciate any feedback on the test code below. Thanks! 

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length, dtype):
        self.len = length
        self.data = torch.randn(length, size).to(dtype)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run(dtype):
    assert dtype in [torch.bfloat16, torch.float16, torch.float32]

    if dtype == torch.bfloat16:
        precision = 'bf16'
    elif dtype == torch.float16:
        precision = '16'
    else:
        precision = '32'

    train_data = DataLoader(RandomDataset(32, 64, dtype), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64, dtype), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64, dtype), batch_size=2)

    model = BoringModel().to(dtype)
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        limit_val_batches=1,
        limit_test_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
        precision=precision, 
        accelerator='gpu',
        devices=2,
        strategy='deepspeed_stage_3'
    )
    checkpoint_path=f'deepspeed_ckpt-{precision}'
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    trainer.save_checkpoint(checkpoint_path)
    trainer.test(model, dataloaders=test_data, ckpt_path=checkpoint_path)


if __name__ == "__main__":
    run(torch.bfloat16)
    run(torch.float16)
    run(torch.float32)
```
